### ISSUE TITLE
Feature Request: Transactions

### ISSUE THREAD

**Issue by Zigur** [balderdashy/waterline#755]
Sorry if I open a new thread, but I could'nt find an old one on the subject...
Are transactions on the Sails/Waterline Roadmap yet? I remembered reading some discussions on them 6 months ago or so, but no updates afterwards. I don't know about the community, but for me at least they would be extremely beneficial (aka sorely needed).
I tried a workaround to use transactions in sails with postgres, using Model.query as it was suggested here:
http://stackoverflow.com/questions/17794784/transactional-sql-with-sails-js 
but I got this issue as that I documented on stackoverflow (no ansers yet):
http://stackoverflow.com/questions/27291992/sails-js-postgres-issue-with-transactions
Has anyone a suggestion / personal workarounds while working with with transactions in Waterline?
Thanks a lot for your help, 
   Massimiliano

**Comment by Batornator** [comment]
I'm also having the issue above (http://stackoverflow.com/questions/27291992/sails-js-postgres-issue-with-transactions). Transactions are currently a must have in our application, any workaround available?

**Comment by Zigur** [comment]
In the end, since transaction support doesn't seem to come soon, I used knex (http://knexjs.org/) to implement the transactional operations I needed. 

https://github.com/biolab-unige/xtens-transact/blob/master/lib/KnexStrategy.js

It works fine, but obviously, you have to access the underlying association tables (if you have associations).

**Comment by prisonier** [comment]
Hello,

the problem with this issue (http://stackoverflow.com/questions/27291992/sails-js-postgres-issue-with-transactions) is that each query is done by a different connexion so transaction is impossible. I've just found this post 10 minutes ago, I did not try it yet but it seems to work.

http://blog.evizija.si/sails-js-waterline-mysql-transactions/

**Comment by elennaro** [comment]
Are you planning to implement transaction somehow soon?

**Comment by dmarcelino** [comment]
@elennaro, it's not slated for `0.11` (next big release) and it's not trivial, check https://github.com/balderdashy/waterline/issues/862#issuecomment-83643605 and https://github.com/balderdashy/waterline/issues/885#issuecomment-93861944. So it may take a few more months. Having said that, any help is much appreciated!

**Comment by prisonier** [comment]
Just to give you my feedback on the link i referenced above (http://blog.evizija.si/sails-js-waterline-mysql-transactions/) -> it is perfectly working for MySQL !

**Comment by elennaro** [comment]
@dmarcelino @prisonier thanks for the answers.

**Comment by jyoungrok** [comment]
@dmarcelino 
Thanks for your comment on the issue(https://github.com/balderdashy/waterline/issues/960#issuecomment-95099001) 
On this link, you suggested that Model.query() method.

Actually, i tried that method for transaction a few months ago. 
Unfortunately, it seems that Model. query() doesn't support multiple queries.
What I mean is that if i implement transaction method like below, it won't work adequately.
Model.query("begin transaction",function(err,result){
     Model.query( another query.....
               Model.query( "rollback" ....
});

because of that, "rollback" method doesn't work......

so as you told me, i can make new connection.
However, i'm worrying if new connection will make some problems (because there will be 2 connections)

**Comment by prisonier** [comment]
@jyoungrok just use the method given in blog i referenced. 

You will setup ONE connection and do everything with this one.

**Comment by jyoungrok** [comment]
@prisonier  Thanks for your comment
i will try that and comment again

**Comment by jyoungrok** [comment]
http://blog.evizija.si/sails-js-waterline-mysql-transactions/
In this blog, i got the news that new release of waterline will involve transaction implementation
Can i know when this implementation will be released?

**Comment by jyoungrok** [comment]
Actually, when i commented, i didn't know about connection pool in detail 
I've caught the point and if i implement like blog @prisonier refer, there won't be much overhead 
because connection pool is just cache of connections for re-use and i will create only one connection and  close it after using that.

Actually, i did it a few months ago but i didn't know in detail
Anyway, thanks for your comment @prisonier 
and finally, i'd like to knwo when the new version will be released :)

**Comment by dmarcelino** [comment]
@jyoungrok  this feature is assigned to [milestone WL_Next](https://github.com/balderdashy/waterline/milestones) which will be released after 0.11, so this should take several months. Having said that, waterline is an open source project developed by the community for the community so you if you want to help implementing this feature your contribution will be appreciated.

**Comment by dmarcelino** [comment]
It's probably a good idea to look at [sails-mysql-transactions](https://github.com/postmanlabs/sails-mysql-transactions#add-transactionid-column-to-all-models) and [knex](http://knexjs.org/#Builder-transacting) when building our own API for transactions, we shouldn't reinvent the wheel.

And this will likely be supported by adapters through a new API... `Transactable`?

**Comment by dmarcelino** [comment]
For reference, there is already a spec for this in #62 written by @mikermcneil:

> > ## tldr;
> 
> 0.5) Add transaction support for uniqueness constraints (already done probably, since we're sending that down as part of the schema)
> 
> 1) Add support for same-adapter transactions, using built-in stuff for the database  (e.g. mysql transactions).  The API looks like `User.transaction().find(....)`, `instance.rollback(cb)` and `instance.commit(cb)`.  If either of those fail, we're in deep trouble.  See the very bottom of this issue.
> 
> 2) Add transaction support (using our new transaction API described above) on top of:
>     + **aggregate** (`createEach()`, plural usage of `update()`, plural usage of `destroy()`)
>     + and **composite** queries (`findOrCreate()`, and `findOrCreateEach()`)
> 
> 3) Add `increment()` and `decrement()`, which are just shortcuts on top of update.  Only thing is they need to be transactional.  Using the same new functionality, of course.
> 
> 4) Figure out how to handle commit() and rollback() errors.  This may already be worked out at the adapter level.
> 
> 5) Add support for `instance.attempts()`, which will reattempt a waterline call *n* times until it works (with incremental back-off).  We can probably leverage async for this.
> 
> 6) Figure out how to do cross-adapter transactions in a user-friendly way.  This is explored in a NOT-user friendly way at the bottom of this issue.

**Comment by atiertant** [comment]
@dmarcelino does this spec are definitive ? is there any `Transactable` adapter interface spec ?

**Comment by dmarcelino** [comment]
No, I don't think so. Nope, at least not yet.

**Comment by atiertant** [comment]
@dmarcelino could be a good starting point ! so if i understood @mikermcneil ,transaction should work like this in waterline : 

    var petTransaction = pet.transaction();
    petTransaction.create()
    petTransaction.find()
    petTransaction.update()
    petTransaction.destroy()
    //async
    if (err) {
      petTransaction.rollback(function (err) { ... });
    }
    else {
      petTransaction.commit(function (err) { ... });
    }

am i correct?
what about populate and cross adapter?
if adapter doesn't support transaction what should waterline do on rollback? (trying to restore last data?)

so adapter need to implements transaction,rollback and commit methods
rollback and commit are easy they just take a callback as argument to handle errors.
transaction i more complex: 
waterline' transaction method should return a transaction specific collection instance ?
so adapter' transaction method should return a transaction specific adapter instance ?

**Comment by devinivy** [comment]
I like that idea as a start.  One issue to think about is how to deal with transactions within callbacks.  Typically in a callback to one of these methods you get a model instance.  Will changes to this model instance implicitly be part of the same transaction?

**Comment by atiertant** [comment]
> Will changes to this model instance implicitly be part of the same transaction?
@devinivy i think it should be the same transaction

the problem with this way is that we can't have a transaction with multiple models...
it could be : 

    var trans = new Transaction();
    pet.transaction(trans).create()
    pet.transaction(trans).find()
    user.transaction(trans).update()
    user.transaction(trans).destroy()
    //async
    if (err) {
      trans.rollback(function (err) { ... });
    }
    else {
      trans.commit(function (err) { ... });
    }

in this way we could choose who is a part of the transaction

**Comment by iyn** [comment]
is there a timeline for transaction support?

**Comment by jyoungrok** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

jyr3357@naver.com 님께 보내신 메일 <Re: [waterline] Feature Request: Transactions (#755)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by nodecode** [comment]
+1 for transaction support

**Comment by formula1** [comment]
TLDR: An Example how to implement it in psuedocode (AKA its not that simple)

Transactions aren't really that simple as this is cross Database. Heres a possible Work around (however probably not the best)

1. First we need to know if there may be a transaction to roll back
  - Specify all documents (model and id) that will be involved in the transaction
  - Specify the properties that will be changed in each

2. We need to know our place in case we do need to roll back
  - This can be done by leaving a watermark on each document "I was here"
  - In the case where a database crashes
    - We see our transaction still exists in the transaction table
    - We go to the first doc and see our watermark is there
    - We go to the next doc and see our watermark is there
    - Etc...
    - We find a doc without a watermark, we continue from there 

3. We need to clean ourselves up after everything is done
  - This can be done by Deleting each watermark and the pending transaction that exists
  - The pending transaction would have to be deleted first before anything happens

4. In the case part of the transaction fails (Which should be done by specifying a query in a way that prevents finding of the document)
  - We Go backward from our current point
    - We remove our watermark and undo the modification we made

This is similar to how MongoDB [explains best practices](http://docs.mongodb.org/manual/tutorial/perform-two-phase-commits/) for two phase commits in their database

Inorder for waterline to have something like this available in an easy to use manner
1) User will need to specify a Model for Transaction
2) User will need to specify Models, arguments to be used in a Query/Validator and operation, Query/Validator for those models and Operation to perform
3) Waterline needs to perform something like the above

Its also possible to ensure that no two transactions modify the same document at the same time
- Before each transaction is made, Check if any other transaction is using our current documents
- If some are found, Find the last added, and set current transaction's "waitingFor" equal to an array of transactions, each being the last 
- In the case a transaction ends - Get all "waitingFor" the ending transaction and remove that dependency. If the dependencies is all resolved for each transaction, resolve the transaction.
- In the case a transaction is canceled 
  - waitingFor = Get all "waitingFor" the canceled transaction
  - canceled.waitingFor = get all documents the document was waiting for
  - For each document the canceled document was waiting for
    - props in common = compare(canceled, canceled.waitingFor[i])
    - For each waitingFor
      - if(waitingFor[i].has(props in common)) waitingFor.push(canceled.waitingFor[i]);
  - For each waitingFor
    - waitingFor[i].waitingFor.remove(canceled)
    - save

**Comment by devinivy** [comment]
@formula1 thanks for thinking about this– I look forward to digging into these suggestions.

**Comment by sailsbot** [comment]
Thanks for posting, @Zigur.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by jyoungrok** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

jyr3357@naver.com 님께 보내신 메일 <Re: [waterline] Feature Request: Transactions (#755)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by elennaro** [comment]
Hey! No way! :D
First bot kills the task we are all waiting for, like I don't know, like a new heart in a hospital!
And then spam? What's going on here :)

**Comment by devinivy** [comment]
This issue should remain open until a PR is made to the roadmap file.

**Comment by kevinburkeshyp** [comment]
fwiw - if you need to implement these in Postgres, here's a library for doing so - https://github.com/Shyp/pg-transactions. 

(If it counts at all I think the approach currently taken by sails-postgresql with respect to unique constraints is great! enforce the constraint at the database level rather than via a transaction means you only need to do 1 database query instead of 4 to validate the constraint. I wrote a little bit more about this here https://kev.inburke.com/kevin/faster-correct-database-queries/)

**Comment by jyoungrok** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

jyr3357@naver.com 님께 보내신 메일 <Re: [waterline] Feature Request: Transactions (#755)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by sailsbot** [comment]
Thanks for posting, @Zigur.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by randallmeeker** [comment]
@sailsbot strikes again

**Comment by particlebanana** [comment]
@randallmeeker thanks, we need to add this to the [Roadmap](https://github.com/balderdashy/waterline/blob/master/ROADMAP.md) so that this can be closed.

**Comment by particlebanana** [comment]
This has been added to the roadmap. There are some interesting projects out there now for doing this including @kevinburkeshyp's project and [Sails MySQL Transactions](https://github.com/postmanlabs/sails-mysql-transactions) from the Postman team.

It's obviously a big challenge and requires bringing connections up to the user level in Waterline. I have some ideas of how this could be done and have some working examples. It should be scheduled for development in the near future but I don't have a timeframe to put on this.
